### PR TITLE
Permissions activity merge

### PR DESCRIPTION
**Pull request by kitsgit**

### CODE REVIEW

File: app/src/main/AndroidManifest.xml [hunk 1]
```diff
@@ -15,6 +15,7 @@
     <uses-permission android:name="android.permission.READ_PHONE_STATE" />
     <uses-permission android:name="android.permission.CHANGE_WIFI_STATE" />
     <uses-permission android:name="android.permission.REQUEST_DELETE_PACKAGES" />
+
```
Comment (Akshay-Kumar-S) [inline app/src/main/AndroidManifest.xml]: unwanted space

File: app/src/main/AndroidManifest.xml [hunk 2]
```diff
@@ -30,10 +31,21 @@
         android:supportsRtl="true"
         android:theme="@style/Theme.AndroidDeviceDetails"
         tools:ignore="AllowBackup">
+        <activity android:name=".ui.BaseAppInfoActivity"
```
Comment (Akshay-Kumar-S) [inline app/src/main/AndroidManifest.xml]: keep same std

File: app/src/main/AndroidManifest.xml [hunk 3]
```diff
@@ -30,10 +31,21 @@
...
+            android:label="Base App Info"
+            android:parentActivityName=".ui.MainActivity"/>
+        <activity
+            android:name=".ui.PermittedAppsActivity"
```
Comment (Akshay-Kumar-S) [inline app/src/main/AndroidManifest.xml]: change name

File: app/src/main/AndroidManifest.xml [hunk 4]
```diff
@@ -30,10 +31,21 @@
...
+            android:label="Permitted Apps"
+            android:parentActivityName=".ui.AppPermissionsActivity"/>
+        <activity
+            android:name=".ui.AppPermissionsActivity"
```
Comment (Akshay-Kumar-S) [inline app/src/main/AndroidManifest.xml:45]: name

File: app/src/main/AndroidManifest.xml [hunk 5]
```diff
@@ -30,10 +31,21 @@
...
+            android:label="App Permissions"
+            android:parentActivityName=".ui.MainActivity" />
         <activity
             android:name=".ui.AppEventActivity"
             android:label="Application Info"
-            android:parentActivityName=".ui.MainActivity">
+            android:parentActivityName=".ui.BaseAppInfoActivity">
```
Comment (Akshay-Kumar-S) [inline app/src/main/AndroidManifest.xml]: only abstract class start with Base

File: app/src/main/java/com/example/androidDeviceDetails/adapters/PermissionsAdapter.kt [hunk 6]
```diff
@@ -0,0 +1,64 @@
...
+import android.view.ViewGroup
+import android.widget.ArrayAdapter
+import com.example.androidDeviceDetails.R
+import com.example.androidDeviceDetails.models.appInfoModels.PermissionsItemViewHolder
+
+class PermissionsAdapter(
+    private var _context: Context,
+    private var resource: Int,
+    private var items: List<String>
+) : ArrayAdapter<String>(_context, resource, items) {
```
Comment (Akshay-Kumar-S) [inline app/src/main/java/com/example/androidDeviceDetails/adapters/PermissionsAdapter.kt:15]: name

File: app/src/main/java/com/example/androidDeviceDetails/adapters/PermissionsAdapter.kt [hunk 7]
```diff
@@ -0,0 +1,64 @@
...
+    }
+
+    @SuppressLint("SetTextI18n")
+    private fun setPermissionsInfoHolder(
+        holder: PermissionsItemViewHolder,
+        position: Int
+    ): PermissionsItemViewHolder {
+        val permission: String = items[position]
+        val drawable = getPermissionDrawable(permission)
+        holder.permissionNameTextView.text = items[position]
```
Comment (Akshay-Kumar-S) [inline app/src/main/java/com/example/androidDeviceDetails/adapters/PermissionsAdapter.kt]: name

File: app/src/main/java/com/example/androidDeviceDetails/adapters/PermissionsAdapter.kt [hunk 8]
```diff
@@ -0,0 +1,64 @@
...
+            )
+            vi.tag = holder
+        } else holder = vi?.tag as PermissionsItemViewHolder
+        setPermissionsInfoHolder(holder, position)
+
+        return vi!!
+    }
+
+    @SuppressLint("SetTextI18n")
+    private fun setPermissionsInfoHolder(
```
Comment (Akshay-Kumar-S) [inline app/src/main/java/com/example/androidDeviceDetails/adapters/PermissionsAdapter.kt]: use?

File: app/src/main/java/com/example/androidDeviceDetails/adapters/PermissionsAdapter.kt [hunk 9]
```diff
@@ -0,0 +1,64 @@
...
+import android.widget.ArrayAdapter
+import com.example.androidDeviceDetails.R
+import com.example.androidDeviceDetails.models.appInfoModels.PermissionsItemViewHolder
+
+class PermissionsAdapter(
+    private var _context: Context,
+    private var resource: Int,
+    private var items: List<String>
+) : ArrayAdapter<String>(_context, resource, items) {
+    @SuppressLint("InflateParams")
```
Comment (Akshay-Kumar-S) [inline app/src/main/java/com/example/androidDeviceDetails/adapters/PermissionsAdapter.kt]: use?

File: app/src/main/java/com/example/androidDeviceDetails/adapters/PermissionsAdapter.kt [hunk 10]
```diff
@@ -0,0 +1,64 @@
...
+        val drawable = getPermissionDrawable(permission)
+        holder.permissionNameTextView.text = items[position]
+        holder.iconView.setImageResource(drawable)
+        return holder
+    }
+
+    private fun getPermissionDrawable(permission: String): Int {
+        var drawable: Int = R.drawable.ic_baseline_android_24
+        try {
+            val permissionInfo: PermissionInfo =
```
Comment (Akshay-Kumar-S) [inline app/src/main/java/com/example/androidDeviceDetails/adapters/PermissionsAdapter.kt]: no need to explicitly mention the type

File: app/src/main/java/com/example/androidDeviceDetails/adapters/PermissionsAdapter.kt [hunk 11]
```diff
@@ -0,0 +1,64 @@
...
+                context.packageManager.getPermissionInfo(permission, 0)
```
Comment (Akshay-Kumar-S) [inline app/src/main/java/com/example/androidDeviceDetails/adapters/PermissionsAdapter.kt]: change 0 to flag name

File: app/src/main/java/com/example/androidDeviceDetails/adapters/PermissionsAdapter.kt [hunk 12]
```diff
@@ -0,0 +1,64 @@
...
+            val groupInfo: PermissionGroupInfo =
+                context.packageManager.getPermissionGroupInfo(permissionInfo.group.toString(), 0)
```
Comment (Akshay-Kumar-S) [inline app/src/main/java/com/example/androidDeviceDetails/adapters/PermissionsAdapter.kt]: from where u got this context?

File: app/src/main/java/com/example/androidDeviceDetails/adapters/PermissionsAdapter.kt [hunk 13]
```diff
@@ -0,0 +1,64 @@
...
+            drawable = groupInfo.icon
+        } catch (e: PackageManager.NameNotFoundException) {
+        } catch (e: Resources.NotFoundException) {
```
Comment (Akshay-Kumar-S) [inline app/src/main/java/com/example/androidDeviceDetails/adapters/PermissionsAdapter.kt]: exception really required?

File: app/src/main/java/com/example/androidDeviceDetails/adapters/PermittedAppsFragmentAdapter.kt [hunk 14]
```diff
@@ -0,0 +1,23 @@
...
+) : FragmentPagerAdapter(fm) {
+
+    override fun getItem(position: Int): Fragment {
+        return when (position) {
+            0 -> PermittedFragment(appList.second)
+            else -> PermittedFragment(appList.first)
+        }
+    }
+
+    override fun getCount() = totalTabs
```
Comment (Akshay-Kumar-S) [inline app/src/main/java/com/example/androidDeviceDetails/adapters/PermittedAppsFragmentAdapter.kt]: make it constant

File: app/src/main/java/com/example/androidDeviceDetails/adapters/PermittedAppsFragmentAdapter.kt [hunk 15]
```diff
@@ -0,0 +1,23 @@
...
+
+import androidx.fragment.app.Fragment
+import androidx.fragment.app.FragmentManager
+import androidx.fragment.app.FragmentPagerAdapter
+import com.example.androidDeviceDetails.fragments.PermittedFragment
+import com.example.androidDeviceDetails.models.permissionsModel.PermittedAppsCookedData
+
+class PermittedAppsFragmentAdapter(
+    fm: FragmentManager, private var totalTabs: Int,
+    private var appList: Pair<List<PermittedAppsCookedData>, List<PermittedAppsCookedData>>
```
Comment (Akshay-Kumar-S) [inline app/src/main/java/com/example/androidDeviceDetails/adapters/PermittedAppsFragmentAdapter.kt]: Pair is not readable

File: app/src/main/java/com/example/androidDeviceDetails/collectors/AppInfoCollector.kt [hunk 16]
```diff
@@ -1,5 +1,4 @@
 package com.example.androidDeviceDetails.collectors
-
```
Comment (Akshay-Kumar-S) [inline app/src/main/java/com/example/androidDeviceDetails/collectors/AppInfoCollector.kt:2]: need space in b/w pkg and import

File: app/src/main/java/com/example/androidDeviceDetails/collectors/PermissionCollector.kt [hunk 17]
```diff
@@ -0,0 +1,70 @@
...
+import com.example.androidDeviceDetails.database.RoomDB
+import com.example.androidDeviceDetails.models.database.AppPermissionsInfo
+import kotlinx.coroutines.GlobalScope
+import kotlinx.coroutines.launch
+
+/**
+ * Implements [BaseCollector].
+ * A time based collector which collects the allowed and denied permissions of individual apps.
+ */
+
```
Comment (Akshay-Kumar-S) [inline app/src/main/java/com/example/androidDeviceDetails/collectors/PermissionCollector.kt]: no space in b/w comment and ()

File: app/src/main/java/com/example/androidDeviceDetails/collectors/PermissionCollector.kt [hunk 18]
```diff
@@ -0,0 +1,70 @@
...
+     * Collect permissions that are allowed and denied for each app using [PackageManager.GET_PERMISSIONS]
+     * store it as a List<[AppPermissionsInfo]>
+     * and writes into [RoomDB.appPermissionDao].
+     */
+
+    override fun collect() {
+        installedApps()
+    }
+
+    private fun installedApps() {
```
Comment (Akshay-Kumar-S) [inline app/src/main/java/com/example/androidDeviceDetails/collectors/PermissionCollector.kt]: name

File: app/src/main/java/com/example/androidDeviceDetails/collectors/PermissionCollector.kt [hunk 19]
```diff
@@ -0,0 +1,70 @@
...
+        val db = RoomDB.getDatabase()!!
+        val appList = context.packageManager.getInstalledPackages(PackageManager.GET_PERMISSIONS)
+        for (packageInfo in appList) {
+            val packageName = packageInfo.packageName.toString()
```
Comment (Akshay-Kumar-S) [inline app/src/main/java/com/example/androidDeviceDetails/collectors/PermissionCollector.kt]: no need .toString()

File: app/src/main/java/com/example/androidDeviceDetails/collectors/PermissionCollector.kt [hunk 20]
```diff
@@ -0,0 +1,70 @@
...
+     */
+
+    override fun collect() {
+        installedApps()
+    }
+
+    private fun installedApps() {
+        val db = RoomDB.getDatabase()!!
+        val appList = context.packageManager.getInstalledPackages(PackageManager.GET_PERMISSIONS)
+        for (packageInfo in appList) {
```
Comment (Akshay-Kumar-S) [inline app/src/main/java/com/example/androidDeviceDetails/collectors/PermissionCollector.kt]: change name

File: app/src/main/java/com/example/androidDeviceDetails/collectors/PermissionCollector.kt [hunk 21]
```diff
@@ -0,0 +1,70 @@
...
+     * store it as a List<List<String>>
+     * and returns it to installedApps().
+     */
+    private fun getPermissions(packageInfo: PackageInfo): List<List<String>> {
+        val allowed: MutableList<String> = ArrayList()
+        val denied: MutableList<String> = ArrayList()
+        val listOfPermissions: MutableList<List<String>> = ArrayList()
+        try {
+            for (i in packageInfo.requestedPermissions.indices) {
+                if (packageInfo.requestedPermissionsFlags[i] and PackageInfo.REQUESTED_PERMISSION_GRANTED != 0) {
```
Comment (Akshay-Kumar-S) [inline app/src/main/java/com/example/androidDeviceDetails/collectors/PermissionCollector.kt]: optimise the code

File: app/src/main/java/com/example/androidDeviceDetails/collectors/PermissionCollector.kt [hunk 22]
```diff
@@ -0,0 +1,70 @@
...
+     * Collects the allowed and denied permission list for each app
+     * store it as a List<List<String>>
+     * and returns it to installedApps().
+     */
+    private fun getPermissions(packageInfo: PackageInfo): List<List<String>> {
+        val allowed: MutableList<String> = ArrayList()
+        val denied: MutableList<String> = ArrayList()
+        val listOfPermissions: MutableList<List<String>> = ArrayList()
+        try {
+            for (i in packageInfo.requestedPermissions.indices) {
```
Comment (Akshay-Kumar-S) [inline app/src/main/java/com/example/androidDeviceDetails/collectors/PermissionCollector.kt]: wrong loop

File: app/src/main/java/com/example/androidDeviceDetails/collectors/PermissionCollector.kt [hunk 23]
```diff
@@ -0,0 +1,70 @@
...
+    }
+
+    private fun installedApps() {
+        val db = RoomDB.getDatabase()!!
+        val appList = context.packageManager.getInstalledPackages(PackageManager.GET_PERMISSIONS)
+        for (packageInfo in appList) {
+            val packageName = packageInfo.packageName.toString()
+            val perms = getPermissions(packageInfo)
+            GlobalScope.launch {
+                val uid = db.appsDao().getIdByName(packageName)
```
Comment (Akshay-Kumar-S) [inline app/src/main/java/com/example/androidDeviceDetails/collectors/PermissionCollector.kt]: dont access db in every loop entry

File: app/src/main/java/com/example/androidDeviceDetails/collectors/PermissionCollector.kt [hunk 24]
```diff
@@ -0,0 +1,70 @@
...
+                val appPermissions =
+                    AppPermissionsInfo(uid, perms[0].toString(), perms[1].toString())
+                try {
+                    db.appPermissionDao().insert(appPermissions)
+                } catch (e: SQLiteConstraintException) {
```
Comment (Akshay-Kumar-S) [inline app/src/main/java/com/example/androidDeviceDetails/collectors/PermissionCollector.kt]: no need to handle exceptions.

File: app/src/main/java/com/example/androidDeviceDetails/controller/ApplicationController.kt [hunk 25]
```diff
@@ -17,7 +17,8 @@ class ApplicationController {
         "AppEventCollector" to AppEventCollector(DeviceDetailsApplication.instance),
         "SignalChangeListener" to SignalChangeCollector(),
         "NetworkUsageCollector" to NetworkUsageCollector(DeviceDetailsApplication.instance),
-        "LocationCollector" to LocationCollector(DeviceDetailsApplication.instance)
+        "LocationCollector" to LocationCollector(DeviceDetailsApplication.instance),
+        "permissions" to PermissionCollector(DeviceDetailsApplication.instance)
```
Comment (Akshay-Kumar-S) [inline app/src/main/java/com/example/androidDeviceDetails/controller/ApplicationController.kt]: P is small letter, naming

File: app/src/main/java/com/example/androidDeviceDetails/cooker/PermissionsCooker.kt [hunk 26]
```diff
@@ -0,0 +1,41 @@
...
+     *>
+     * Overrides [cook] in [BaseCooker]
+     * >
+     * @param callback A callback that accepts the cooked list once the cooking is done.
+     */
+
+    @Suppress("UNCHECKED_CAST")
+    override fun <T> cook(time: TimePeriod, callback: ICookingDone<T>) {
+        GlobalScope.launch(Dispatchers.IO) {
+            var listOfPermissions: MutableList<String> = ArrayList()
```
Comment (Akshay-Kumar-S) [inline app/src/main/java/com/example/androidDeviceDetails/cooker/PermissionsCooker.kt]: unwanted casting

File: app/src/main/java/com/example/androidDeviceDetails/cooker/PermissionsCooker.kt [hunk 27]
```diff
@@ -0,0 +1,41 @@
...
+            val db = RoomDB.getDatabase()!!
+            val appList = db.appPermissionDao().getPermittedApps()
```
Comment (Akshay-Kumar-S) [inline app/src/main/java/com/example/androidDeviceDetails/cooker/PermissionsCooker.kt]: getPermittedApps change name

**Review by richardjoy530 — COMMENTED** [review]
Reviewed

**Pull request merged**